lib/softirq.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2019, Xilinx Inc. and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <errno.h>
+#include <metal/atomic.h>
+#include <metal/irq.h>
+#include <metal/softirq.h>
+#include <metal/utilities.h>
+#include <string.h>
+
+#define METAL_SOFTIRQ_NUM 64
+
+#define METAL_SOFTIRQ_ARRAY_DECLARE(num) \
+	static const int metal_softirq_num = num; \
+	static struct metal_irq metal_softirqs[num]; \
+	static unsigned long metal_softirq_assigned[metal_bitmap_longs(num)]; \
+	static unsigned long metal_softirq_pending[metal_bitmap_longs(num)]; \
+	static unsigned long metal_softirq_enabled[metal_bitmap_longs(num)]; \
+
+METAL_SOFTIRQ_ARRAY_DECLARE(METAL_SOFTIRQ_NUM)
+
+static void
+metal_softirq_set_state(struct metal_irq_controller *irq_cntr,
+			int irq, unsigned int state);
+
+static METAL_IRQ_CONTROLLER_DECLARE(metal_softirq_cntr,
+				    METAL_IRQ_ANY, 0,
+				    NULL,
+				    metal_softirq_set_state,
+				    NULL)
+
+static atomic_flag metal_softirq_lock;
+
+static void metal_softirq_set_enable(int bit, unsigned int state)
+{
+
+	while(atomic_flag_test_and_set(&metal_softirq_lock) != 0);
+	if (state ==  METAL_IRQ_ENABLE) {
+		metal_bitmap_set_bit(metal_softirq_enabled, bit);
+	} else {
+		metal_bitmap_clear_bit(metal_softirq_enabled, bit);
+	}
+	atomic_flag_clear(&metal_softirq_lock);
+
+}
+
+static void metal_softirq_set_pending(int bit)
+{
+
+	while(atomic_flag_test_and_set(&metal_softirq_lock) != 0);
+	metal_bitmap_set_bit(metal_softirq_pending, bit);
+	atomic_flag_clear(&metal_softirq_lock);
+
+}
+
+static void
+metal_softirq_set_state(struct metal_irq_controller *irq_cntr,
+			int irq, unsigned int state)
+{
+	int bit;
+
+	(void)irq_cntr;
+	bit = irq - metal_softirq_cntr.irq_base;
+	if (bit < 0) {
+		return;
+	}
+	if (state == METAL_IRQ_PENDING) {
+		metal_softirq_set_pending(bit);
+	} else if (state == METAL_IRQ_ENABLE ||
+		   state == METAL_IRQ_DISABLE) {
+		metal_softirq_set_enable(bit, state);
+	} else {
+		return;
+	}
+}
+
+int metal_softirq_init()
+{
+	int irq_base;
+	int irq_num = metal_softirq_num;
+
+	irq_base = metal_irq_allocate(METAL_IRQ_ANY, irq_num);
+	if (irq_base < 0)
+		return -EINVAL;
+	metal_irq_controller_set_range(&metal_softirq_cntr,
+				       irq_base, irq_num);
+	metal_irq_controller_set_irqs(&metal_softirq_cntr, metal_softirqs);
+	return 0;
+}
+
+int metal_softirq_allocate(int num)
```

Comment:
I think we should require softirq allocation to be done before multithreading is started.  This way we don't have to do any locking.  

Do we need a bitmap to keep track of allocated softirq numbers?  Isn't it enough to just have a counter that is incremented for every allocation?

Reply:
Right, as unlike hard IRQ, there is no predefined soft IRQs, so there should be no holes between the allocated soft IRQs, can remove the bitmap.